### PR TITLE
HPCC-28143 Loosen WsStore SDS Write Lock Access

### PR DESCRIPTION
**Pull request by rpastrana**
- Do not request write lock on reads

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@jakesmith as discussed offline, not sure this change will make any sort of impact, but it should lessen the chance of contention due to held SDS locks.

Performed several load tests consisting of 15 continuous threads requesting set operations, coupled with various read/write/delete requests.

File: esp/services/ws_store/espstorelib/daliKVStore.cpp [hunk 1]
```diff
@@ -166,7 +166,7 @@ IPropertyTree * CDALIKVStore::getAllKeyProperties(const char * storename, const
 
     xpath.appendf("/%s/%s", ns, key);
 
-    Owned<IRemoteConnection> conn = querySDS().connect(xpath.str(), myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT_KVSTORE);
+    Owned<IRemoteConnection> conn = querySDS().connect(xpath.str(), myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_KVSTORE);
     if (!conn)
         throw MakeStringException(-1, "DALI Keystore fetchKeyProperties(): Unable to connect to DALI KeyValue store path '%s'", xpath.str()); //rodrigo, not sure if this is too much info
```
Comment (jakesmith) [inline esp/services/ws_store/espstorelib/daliKVStore.cpp:172]: not related, but :
```
    Owned<IPropertyTree> keytree = conn->getRoot();
    if (!keytree.get())
        throw MakeStringException(-1, "DALI KV Store fetchKeyProperties(): Unable to access key '%s'", key); //this store doesn't exist
```

isn't necessary. If a connection succeeds, it has a root by definition. Also not necessary to link the root, i.e. would be more efficient/conventional to do :
```
    return conn->getRoot();
```

as getPropTree(".") is just self, so same as root.

**Review by jakesmith — APPROVED** [review]

**Comment by jakesmith** [comment]
> @jakesmith as discussed offline, not sure this change will make any sort of impact, but it should lessen the chance of contention due to held SDS locks.
> 
> Performed several load tests consisting of 15 continuous threads requesting set operations, coupled with various read/write/delete requests.

@rpastrana - approved look good, 1 comment, but unrelated to changes.

It would be good to add that load test or similar as a unittest?

**Review by jakesmith — APPROVED** [review]
@rpastrana - approving again (but needs squashing),

however:
https://github.com/hpcc-systems/HPCC-Platform/pull/16482#discussion_r956110789

that one was an example, which is now changed, but there are a quite a few others that were falling the same pattern of getting and linking root, checking it is not null, throwing an exception. Now you've removed the unecessary code for 1, you might want to look at the rest (search for conn->getRoot())

File: esp/services/ws_store/espstorelib/daliKVStore.cpp [hunk 2]
```diff
@@ -166,15 +166,11 @@ IPropertyTree * CDALIKVStore::getAllKeyProperties(const char * storename, const
...
 
-    Owned<IPropertyTree> keytree = conn->getRoot();
-    if (!keytree.get())
-        throw MakeStringException(-1, "DALI KV Store fetchKeyProperties(): Unable to access key '%s'", key); //this store doesn't exist
-
-    return(keytree->getPropTree("."));
+    return(conn->getRoot());
```
Comment (jakesmith) [inline esp/services/ws_store/espstorelib/daliKVStore.cpp:173]: trivial: unusual to add brackets around return result.

**Comment by rpastrana** [comment]
@jakesmith squashed
@richardkchapman ready to merge

**Pull request merged**